lib/client/pool_e2e_test.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2016 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Test cases (Pool object e2e):
+// Test 1: Pass two fully populated PoolSpecs and expect the series of operations to succeed.
+// Test 2: Pass one partially populated PoolSpec and another fully populated PoolSpec and expect the series of operations to succeed.
+// Test 3: Pass one fully populated PoolSpec and another empty PoolSpec and expect the series of operations to succeed.
+// Test 4: Pass two fully populated PoolSpecs with two PoolMetadata (one IPv4 and another IPv6) and expect the series of operations to succeed.
+
+// Series of operations each test goes through:
+// Update meta1 - check for failure (because it doesn't exist).
+// Create meta1 with spec1.
+// Apply meta2 with spec2.
+// Get meta1 and meta2, compare spec1 and spec2.
+// Update meta1 with spec2.
+// Get meta1 compare spec2.
+// List (empty Meta) ... Get meta1 and meta2.
+// List (using Meta1) ... Get meta1.
+// Delete meta1.
+// Get meta1 ... fail.
+// Delete meta2.
+// List (empty Meta) ... Get no entries (should not error).
+
+package client_test
+
+import (
+	"errors"
+	"log"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+
+	"github.com/projectcalico/libcalico-go/lib/api"
+	"github.com/projectcalico/libcalico-go/lib/testutils"
+)
+
+var _ = Describe("Pool tests", func() {
+
+	DescribeTable("Pool e2e tests",
+		func(meta1, meta2 api.PoolMetadata, spec1, spec2 api.PoolSpec) {
+
+			// Erase etcd clean.
+			testutils.CleanEtcd()
+
+			// Create a new client.
+			c, err := testutils.NewClient("")
+			if err != nil {
+				log.Println("Error creating client:", err)
+			}
+			By("Updating the pool before it is created")
+			_, outError := c.Pools().Update(&api.Pool{Metadata: meta1, Spec: spec1})
+
+			// Should return an error.
+			Expect(outError.Error()).To(Equal(errors.New("resource does not exist: Pool(cidr=10.0.0.0/24)").Error()))
+
+			By("Create, Apply, Get and compare")
+
+			// Create a pool with meta1 and spec1.
+			_, outError = c.Pools().Create(&api.Pool{Metadata: meta1, Spec: spec1})
+
+			// Apply a pool with meta2 and spec2.
+			_, outError = c.Pools().Apply(&api.Pool{Metadata: meta2, Spec: spec2})
+
+			// Get pool with meta1.
+			outPool1, outError1 := c.Pools().Get(meta1)
+			log.Println("Out Pool object: ", outPool1)
```

Comment:
>We should really be checking the error code after each of these calls - much like we would for code.

What difference does it make if I assert the error right after vs in line 85? I think keeping them together makes it easier to understand each step we are taking for the FVs with a comment explaining what the expected result is for each step.